lib/pg_eventstore/client.rb

```diff
@@ -0,0 +1,133 @@
+    #   PgEventstore to return the original event instead a link event.
+    # @option options [Hash] :filter provide it to filter events. You can filter by: stream and by event type. Filtering
+    #   by stream is only available when reading from "all" stream.
+    #   Examples:
+    #     # Filtering by stream's context. This will return all events which #context is 'User
+    #     PgEventstore.client.read(
+    #       PgEventstore::Stream.all_stream,
+    #       options: { filter: { streams: [{ context: 'User' }] } }
+    #     )
+    #
+    #     # Filtering by several stream's contexts. This will return all events which #context is either 'User' or
+    #     # 'Profile'
+    #     PgEventstore.client.read(
+    #       PgEventstore::Stream.all_stream,
+    #       options: { filter: { streams: [{ context: 'User' }, { context: 'Profile' }] } }
+    #     )
+    #
+    #     # Filtering by a mix of specific stream and a context. This will return all events which #context is 'User' or
+    #     # events belonging to the stream with { context: 'Profile', stream_name: 'ProfileFields', stream_id: '123' }
+    #     PgEventstore.client.read(
+    #       PgEventstore::Stream.all_stream,
+    #       options: {
+    #         filter: {
+    #           streams: [
+    #             { context: 'User' },
+    #             { context: 'Profile', stream_name: 'ProfileFields', stream_id: '123' }
+    #           ]
+    #         }
+    #       }
+    #     )
+    #
+    #     # Filtering the a mix of context and event type
+    #     PgEventstore.client.read(
+    #       PgEventstore::Stream.all_stream,
+    #       options: { filter: { streams: [{ context: 'User' }], event_types: ['MyAwesomeEvent'] } }
+    #     )
+    #
+    #     # Filtering by specific event when reading from the specific stream
+    #     PgEventstore.client.read(stream, options: { filter: { event_types: ['MyAwesomeEvent'] } })
+    # @param skip_middlewares [Boolean] whether to skip middlewares. Defaults to false
+    # @return [Array<PgEventstore::Event>]
+    # @raise [PgEventstore::StreamNotFoundError]
+    def read(stream, options: {}, skip_middlewares: false)
+      Commands::Read.
+        new(queries(middlewares(skip_middlewares))).
+        call(stream, options: { max_count: config.max_count }.merge(options))
+    end
+
+    private
+
+    # @param skip_middlewares [Boolean]
+    # @return [Array<Object<#serialize, #deserialize>>]
+    def middlewares(skip_middlewares = false)
+      return config.middlewares unless skip_middlewares
+
+      []
+    end
+
+    # @return [PgEventstore::Connection]
+    def connection
+      PgEventstore.connection(config.name)
+    end
+
+    # @param middlewares [Array<Object<#serialize, #deserialize>>]
+    # @return [PgEventstore::Queries]
+    def queries(middlewares)
+      Queries.new(
+        connection,
+        EventSerializer.new(middlewares),
+        PgresultDeserializer.new(middlewares, config.event_class_resolver)
```

Comment:
Should be `PgResultDeserializer` I believe?

Reply:
Probably yes. Initialize it is not camel-cased, because it is how it is named in `pg` gem. But, probably, in this place it is more meaningful to name it properly